crates/ra_hir/src/traits.rs

```diff
@@ -49,4 +50,34 @@ pub enum TraitItem {
     TypeAlias(TypeAlias),
     // Existential
 }
+// FIXME: not every function, ... is actually a trait item. maybe we should make
+// sure that you can only turn actual trait items into TraitItems. This would
+// require not implementing From, and instead having some checked way of
+// casting them.
```

Comment:
Good point!

A whole lot of things would love access to `db` at some point, and that makes traits much less useful. I wonder if we should have

```rust
struct WithDb<DB, T> {
    db: DB,
    value: T
}
```

and implement From's for `WithDb<impl DefDatabase, Function>`

Reply:
I think `From` still wouldn't work, since the point of this would be that the conversion would be fallible ;)

Reply:
TryFrom is stable on Thursday ;)